### PR TITLE
campaigns: Add webhook tests for Bitbucket

### PR DESCRIPTION
**Pull request by ryanslade**
Closes: https://github.com/sourcegraph/sourcegraph/issues/9104

### CODE REVIEW

File: enterprise/internal/campaigns/webhooks_test.go [hunk 1]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
 }
 
+// Run from integration_test.go
+func testBitbucketWebhook(db *sql.DB) func(*testing.T) {
+	return func(t *testing.T) {
+		now := time.Now()
+		clock := func() time.Time {
+			return now.UTC().Truncate(time.Microsecond)
+		}
+		now = clock()
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:203]: I might be missing something, but... is this the idea?

```suggestion
		now := time.Now().UTC().Truncate(time.Microsecond)
		clock := func() time.Time { return now }
```

File: enterprise/internal/campaigns/webhooks_test.go [hunk 2]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
+				continue
+			}
+			if result.Repo.Name == "bitbucket.sgdev.org/SOUR/automation-testing" {
+				bitbucketRepo = result.Repo
+			}
+		}
+
+		if bitbucketRepo == nil {
+			t.Fatal("repo not found")
+		}
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:264]: Maybe we should think about adding a `GetRepo` method to `BitbucketServerSource`. The [Github source already has one](https://github.com/sourcegraph/sourcegraph/blob/3906e0f12edc77d6cfa1d66c7e3b138ebae5ad4a/cmd/repo-updater/repos/github.go#L286-L294). And we've also used that in tests.

But the "is archived" thing might be a blocker?


```go
// cmd/repo-updater/repos/bitbucketserver.go

func (s *BitbucketServerSource) GetRepo(ctx context.Context, projectKey, repoSlug string) (*Repo, error) {
	repo, err := s.client.Repo(ctx, projectKey, repoSlug)
	if err != nil {
		return nil, err
	}
	// TODO: do we need to fetch all archived labels?
	isArchived := false
	return s.makeRepo(repo, isArchived), nil
}
```

File: enterprise/internal/campaigns/webhooks_test.go [hunk 3]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
+				CampaignIDs:         []int64{campaign.ID},
+			},
+		}
+
+		err = store.CreateChangesets(ctx, changesets...)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = SyncChangesets(ctx, repoStore, store, cf, changesets...)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:305]: Nice. I know it's costly, but I do like how integration-y these tests are.

File: enterprise/internal/campaigns/webhooks_test.go [hunk 4]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
+				tc := loadWebhookTestCase(t, fixtureFile)
+
+				// Send all events twice to ensure we are idempotent
+				for i := 0; i < 2; i++ {
+					for _, event := range tc.Payloads {
+						req, err := http.NewRequest("POST", "", bytes.NewReader(event.Data))
+						if err != nil {
+							t.Fatal(err)
+						}
+						req.Header.Set("X-Hub-Signature", sign(t, event.Data, []byte(secret)))
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:339]: Don't you also need to set `X-Event-Key` to the `payload_type` from the fixture?

See this code here:

https://github.com/sourcegraph/sourcegraph/blob/3906e0f12edc77d6cfa1d66c7e3b138ebae5ad4a/internal/extsvc/bitbucketserver/events.go#L11-L30

File: enterprise/internal/campaigns/webhooks_test.go [hunk 5]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				tc := loadWebhookTestCase(t, fixtureFile)
+
+				// Send all events twice to ensure we are idempotent
+				for i := 0; i < 2; i++ {
+					for _, event := range tc.Payloads {
+						req, err := http.NewRequest("POST", "", bytes.NewReader(event.Data))
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:335]: What about the `?externalServiceID` query param?

File: enterprise/internal/campaigns/webhooks_test.go [hunk 6]
```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
...
+						hook.ServeHTTP(rec, req)
+						resp := rec.Result()
+
+						if resp.StatusCode != http.StatusOK {
+							t.Fatalf("Non 200 code: %v", resp.StatusCode)
+						}
+					}
+				}
+
+				have, _, err := store.ListChangesetEvents(ctx, ListChangesetEventsOpts{Limit: 1000})
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks_test.go:351]: I know, I know, but... let's just be safe 😎 

```suggestion
				have, _, err := store.ListChangesetEvents(ctx, ListChangesetEventsOpts{Limit: -1})
```

File: enterprise/internal/campaigns/webhooks_test.go (hunk 2 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/webhooks_test.go:264]: If not, I'd suggest extracting the whole "get a single repo from source" into a small helper function like this. Just takes some "noise" out of the test:

https://github.com/sourcegraph/sourcegraph/blob/3906e0f12edc77d6cfa1d66c7e3b138ebae5ad4a/enterprise/internal/campaigns/resolvers/resolver_test.go#L1512-L1530

File: enterprise/internal/campaigns/testdata/fixtures/webhooks/bitbucketserver/add-review.json [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+                },
+                "avatarUrl": "/users/milton/avatar.png?s=64\u0026v=1576525349000"
+              }
+            ],
+            "removedReviewers": []
+          }
+        }
+      },
+      {
+        "payload_type": "",
```
Comment (mrnugget) [inline enterprise/internal/campaigns/testdata/fixtures/webhooks/bitbucketserver/add-review.json:221]: It's a bit confusing seeing this empty string here, so I'd suggest either omitting it or adding a dummy value, maybe:

```suggestion
        "payload_type": "IGNORED_FOR_THIS_EVENT",
```

File: enterprise/internal/campaigns/webhooks_test.go (hunk 1 above)
Comment (ryanslade) [reply enterprise/internal/campaigns/webhooks_test.go:203]: I'll clean that up, it was copied from setup in some old tests and doesn't really make sense anymore

**Pull request merged**